src/debug/debugService.ts

```diff
@@ -0,0 +1,273 @@
+            vscode.window.showErrorMessage("Cannot resolve debug/application port from Dockerfile.");
+            return;
+        }
+
+        vscode.window.withProgress({ location: vscode.ProgressLocation.Window }, async (p) => {
+            let appName;
+            try {
+                // Build/push docker image.
+                p.report({ message: "Building docker image..."});
+                const shellOpts = Object.assign({ }, shell.execOpts(), { cwd });
+                const image = await this.buildDockerImageStep(imagePrefix, shellOpts);
+
+                // Run docker image in k8s container.
+                p.report({ message: "Running docker image on Kubernetes..."});
+                appName = await this.runAsDeploymentStep(image, [portInfo.appPort, portInfo.debugPort], containerEnv);
+
+
+                // Find the running debug pod.
+                p.report({ message: "Finding the debug pod..."});
+                const podName = await this.findDebugPodStep(appName);
+                
+                // Wait for the debug pod status to be running.
+                p.report({ message: "Waiting for the pod to be ready..."});
+                await this.waitForPodStep(podName);
+    
+                // Setup port-forward.
+                p.report({ message: "Setting up port forwarding..."});
+                const proxyResult = await this.setupPortForwardStep(podName, portInfo.debugPort, portInfo.appPort);
+                
+                // Start debug session.
+                p.report({ message: `Starting ${this.debugProvider.getDebuggerType()} debug session...`});
+                await this.startDebugSessionStep(appName, cwd, proxyResult);
+            } catch (error) {
+                vscode.window.showErrorMessage(error);
+                kubeChannel.showOutput(`Debug on Kubernetes failed. The errors was: ${error}.`);
+                if (appName) {
+                    await this.cleanupResourceStep(`deployment/${appName}`);
+                }
+            }
+            return null;
+        });
+    }
+
+    /**
+     * In attach mode, the user should choose the running pod first, then the debugger will attach to it.
+     * 
+     * @param workspaceFolder the active workspace folder.
+     * @param pod the debug pod name.
+     */
+    public async attachDebug(workspaceFolder: vscode.WorkspaceFolder, pod?: string): Promise<void> {
+        // TODO
+        return;
+    }
+
+    private async buildDockerImageStep(imagePrefix: string, shellOpts: any): Promise<string> {
+        kubeChannel.showOutput("Starting to build/push docker image...", "Docker build/push");
+        if (!imagePrefix) {
+            // In order to allow local kubernetes cluster (e.g. minikube) to have access to local docker images,
+            // need override docker-env before running docker build.
+            const dockerEnv = await dockerUtils.resolveDockerEnv(this.kubectl);

```

Comment:
It does the same thing as `minikube docker-env`. It's used to get docker env of the kubernetes cluster, and set it to the shell context of running `docker build` via local `docker` client so that the built image is generated in the kubernetes docker directly. 

   This just worked for local `minikube`. For remote kubernetes cluster, the user need specify `imagePrefix` via `vsdocker.imageUser` settings and pushed to `docker hub` or `ACR`.